### PR TITLE
Lots of comments

### PR DESCRIPTION
**Pull request by Takareer**
Please run the unit tests on this! I don't have access to a linux computer so I cannot do it myself.

### CODE REVIEW

File: units_cpp/action.cc [hunk 1]
```diff
@@ -2,9 +2,9 @@
 #include "colour.h"
 
 
-//----------------------------------------------------------------------------------------------------------------------
+//------------------------------------------------------------------------------
```
Comment (neynt) [inline units_cpp/action.cc:5]: Much thanks!

File: units_cpp/game_state.h [hunk 2]
```diff
@@ -1,42 +1,56 @@
...
-  int players_ready;
-
-  void begin_turn();
-  void execute_battle(); // Executes the battle and advances the turn
+	/* Contains all information relevant to a single Adrestia game.
+	 * Crucially, (am able to reconstruct EVERYTHING about a game) <=> (can construct/have a GameState for the game)
+	 */
+	public:
+		GameState(const GameRules &rules, int num_players);
+		GameState(const GameRules &rules, const json &j);  // From GameRules and a json version of a GameState
```
Comment (neynt) [inline units_cpp/game_state.h]: redundant comment; it contains no information that you can't glean from reading the types!

File: units_cpp/game_state.h [hunk 3]
```diff
@@ -1,42 +1,56 @@
...
+		GameState(const GameState &game_state);
+		GameState(const GameView &view, const std::vector<Tech> &techs);  // Determinization with the given hidden information
```
Comment (neynt) [inline units_cpp/game_state.h]: image

File: units_cpp/game_state.h [hunk 4]
```diff
@@ -1,42 +1,56 @@
...
+
+		// Returns false if the action is illegal; true otherwise.
```
Comment (neynt) [inline units_cpp/game_state.h]: In general, replacing "false if bad; true otherwise" with "true iff good" or even "checks whether good" enhances readability.

File: units_cpp/game_state.h [hunk 5]
```diff
@@ -1,42 +1,56 @@
+/* Defines the class GameState, which contains all information relevant to a single Adrestia game. */
+
```
Comment (neynt) [inline units_cpp/game_state.h]: I disagree with two blank lines above the #pragma once (really, any sort of double-blank-lining), though getting into a fight about it will probably cause more productivity loss than simply acquiescing to your taste in code style.

File: units_cpp/game_state.h [hunk 6]
```diff
@@ -1,42 +1,56 @@
...
+
 #pragma once
-#include <vector>
+
 #include <memory>
 #include <random>
+#include <vector>
+
```
Comment (neynt) [inline units_cpp/game_state.h:10]: Though -- thanks for separating STL imports vs. our own.

File: units_cpp/game_state.h [hunk 7]
```diff
@@ -1,42 +1,56 @@
...
+		// Returns false if the action is illegal; true otherwise.
+		//     In the case of an illegal action, nothing is changed.
+		bool perform_action(int player, const Action &action);
+
+		void get_view(GameView &view, int player) const;
+		std::vector<int> get_winners() const;  // 1 Winner = normal, all players = draw, no winners = game not done
+		const GameRules &get_rules() const;
+		int get_turn() const;  // Get current turn. Turn 0 means uninitialized; turn 1 is the first 'playable' turn.
+		const std::vector<Player> &get_players() const;
+		const std::vector<std::vector<std::vector<Action>>> &get_action_log() const;  // [turn][player][action]
```
Comment (neynt) [inline units_cpp/game_state.h]: I'm gonna go ahead and move some of these comments to the line above, as I am a cranky boy who enjoys long lines only when I write them.

File: units_cpp/game_view.cc [hunk 8]
```diff
@@ -1,57 +1,69 @@
...
-  }
-  generate_build_units_actions(actions, units, coins, max_units, next_it, end);
+
+void GameView::generate_build_units_actions(std::vector<Action> &actions,  // Possible actions (accumulator)
+                                            std::vector<std::string> &units,
+                                            int coins,
+                                            int max_units,
+                                            std::map<std::string, UnitKind>::const_iterator begin,
+                                            std::map<std::string, UnitKind>::const_iterator end
+                                           ) const {
```
Comment (neynt) [inline units_cpp/game_view.cc]: I'm gonna create a newline after the ( and indent using a single tab instead of a billion spaces. In general, I think allowing any sort of variable-width alignment is quite bad precisely because stuff like this happens. (Though here, I do this mainly because starting ANY line with spaces confuses vim-sleuth into making tabs 8-wide.)

File: units_cpp/game_view.cc [hunk 9]
```diff
@@ -1,57 +1,69 @@
...
-    std::vector<Action> actions;
-    std::vector<std::string> units;
-    generate_build_units_actions(actions, units, view_player.coins, rules->get_unit_cap() - view_player.units.size(),
-        rules->get_unit_kinds().begin(), rules->get_unit_kinds().end());
-    return actions;
-  } else {
-    return std::vector<Action>();
-  }
+	/* Generate a list of ALL legal actions for this View's Player.
+	 *     The result will be empty if the game is over, but also if we are waiting for other players to make their
```
Comment (neynt) [inline units_cpp/game_view.cc]: In vim, you can reflow text to 80 characters wide by selecting it visually and using gq. I encourage you to find and use a similar feature in your editor. (gq is particularly cool in that it also inserts/removes comment characters where needed).

peek 2018-09-23 00-26

Separately, it turns out "but also if" is the same as "or" -- a useful tip for concision.

File: units_cpp/player.cc [hunk 10]
```diff
@@ -1,65 +1,75 @@
...
-    , alive(player.alive)
-    , coins(player.coins)
-    , tech(player.tech)
-    , next_unit(player.next_unit) {}
+Player::Player(const Player &player):
+		units(player.units),
+		alive(player.alive),
+		coins(player.coins),
+		tech(player.tech),
+		next_unit(player.next_unit)
```
Comment (neynt) [inline units_cpp/player.cc]: Comma-first style is an established thing that Charles and I have used and grown somewhat fond of. It may seem ugly at first glance from your experience with the comma in English writing, but it has distinct advantages in programming. See this gist for some of the motivation behind it: https://gist.github.com/isaacs/357981

I'll be changing this back.

File: units_cpp/player.cc [hunk 11]
```diff
@@ -1,65 +1,75 @@
...
-    , coins(j["coins"])
-    , tech(j["tech"])
-    , next_unit(j["next_unit"]) {
-  for (auto it = j["units"].begin(); it != j["units"].end(); it++) {
-    const UnitKind &kind = rules.get_unit_kind(it.value()["kind"]);
-    units.emplace(std::stoi(it.key()), Unit(kind, it.value()));
-  }
+Player::Player(const GameRules &rules, const json &j): alive(j["alive"]),
+                                                       coins(j["coins"]),
+                                                       tech(j["tech"]),
```
Comment (neynt) [inline units_cpp/player.cc]: ... especially since this other style causes this to happen

File: units_cpp/tech.cc [hunk 12]
```diff
@@ -1,75 +1,95 @@
...
 #include <sstream>
-#include "tech.h"
+
 #include "json.h"
+#include "tech.h"
+
+
+Tech::Tech():
+		red(0),
+		green(0),
```
Comment (neynt) [inline units_cpp/tech.cc]: If it fits on one line, there's no need to make it five (unless there is some very advantageous gain in code alignment or readability; not the case here).

File: units_cpp/game_state.cc [hunk 13]
```diff
@@ -1,240 +1,350 @@
...
+}
+
+
+bool GameState::perform_action(int player, const Action &action) {
+	if (get_winners().size() > 0) {
+		// Game is already over - all actions illegal
+		return false;
+	}
+	if ((size_t)player >= players.size()) {
+		// Nonexistent player
```
Comment (neynt) [inline units_cpp/game_state.cc]: Thanks for interpreting all these checks in plain English! I admit I was getting a bit tired of re-understanding them each time myself. I think splitting the single-line checks into three lines w/ an open/closing curly brace is a bit overkill and I will be reverting those, but I'll keep all your comments.

File: units_cpp/tech.h [hunk 14]
```diff
@@ -1,26 +1,31 @@
...
-  void add(const Tech &t);
-  void max(const Tech &t);
-  void increment(Colour colour);
-  bool includes(const Tech &t) const;
+		Tech();
+		Tech(int red, int green, int blue);
+		void add(const Tech &t);
+		void max(const Tech &t);  // Sets all values of this tech to the max of (this tech, the given tech)
+		void increment(Colour colour);  // Increment this colour by one. Will not increment BLACK.
+		bool includes(const Tech &t) const;  // Check if this tech is a proper superset of the given one.
```
Comment (neynt) [inline units_cpp/tech.h]: You probably just mean "superset". "proper superset" means "superset AND not equal to".

**Comment by neynt** [comment]
Thanks for this. Ran unit tests and they passed. I took a look through and made changes I would suggest, liberally leaving github comments to explain my thought process.

In addition, I think it's worth distinguishing between comments that kind of represent your stream-of-consciousness while reading code, vs. comments that actually explain surprising behavior or cut to a deep intuition. The second type of comment tends to be far more useful than the first.

**Review by neynt — APPROVED** [review]

File: units_cpp/game_view.cc (hunk 8 above)
Comment (Takareer) [reply units_cpp/game_view.cc]: Okay!
